aten/src/ATen/native/cudnn/Conv.cpp

```diff
@@ -827,39 +829,76 @@ void raw_cudnn_convolution_forward_out(
     &zero, args.odesc.desc(), output.data_ptr()));
 }
 
+// Precondition: layout is either NCF or NFC.
+Tensor allocate_output(at::IntList sizes, Layout layout, const TensorOptions& options) {
+  switch (layout) {
+    case Layout::NCF:
+      return at::empty(sizes, options);
+    case Layout::NFC: {
+      // Allocate contiguous output with sizes in NFC layout
+      AT_ASSERT(sizes.size() >= 3);
+      std::array<int64_t, max_all_dims> transposed_sizes;
+      transposed_sizes[0] = sizes[0];
+      for (int64_t i = 2; i < sizes.size(); ++i) {
+        transposed_sizes[i - 1] = sizes[i];
+      }
+      transposed_sizes[sizes.size() - 1] = sizes[1];
+      auto output = at::empty(ArrayRef<int64_t>(transposed_sizes).slice(0, sizes.size()), options);
+
+      // Permute the dimensions to fix sizes to NCF
+      return output.permute(LayoutPermutation{Layout::NFC, Layout::NCF, static_cast<int64_t>(sizes.size())});
+    }
+  }
+  AT_ASSERTM(false, "Unhandled layout in allocate_output");
+}
+
+Layout get_input_layout(TensorArg* input_arg) {
+  auto & input = input_arg->tensor;
+  if (auto layout = getLayout(input)) {
+    return *layout;
+  }
+  // For GPUs with WMMA (Tensor cores) it's usually faster to use NFC if
+  // we're paying the cost of a transpose anyway.
```

Comment:
Yes, it will transpose the weights, but it's not a mutable change. It's done transparently in every call. From what I saw in my benchmarks the weight transpose cost is tiny compared to the perf benefits, so it seems like a reasonable thing to do.

Reply:
I'm not sure - suppose you have non-contiguous NCF input for whatever reason, e.g. you operate on just some channels - you don't expect the output to suddenly become NFC.

Reply:
Sure, I'm definitely not saying that this gives great usability to people who want to start using NCF in the eager frontend, since you can easily mess up by passing in inputs of slightly different strides. But this is enough to e.g. experiment with JIT layout rewrites.

But there's no such thing as non-contiguous NCF. There's either NFC (contiguous), NCF (always non-contiguous) and then all other kinds of non-contiguous tensors.